File: scripts/compile_videos.js
#!/usr/bin/env node

/**
 * Script Node.js de compilation de vidéos pour n8n
 * Compile 8 vidéos de 8 secondes en une seule vidéo finale
 */

const fs = require('fs').promises;
const path = require('path');
const { exec } = require('child_process');
const { promisify } = require('util');

const execAsync = promisify(exec);

/**
 * Configuration des plateformes
 */
const PLATFORM_CONFIGS = {
  tiktok: {
    size: '1080:1920',
    bitrate: '3000k',
    fps: 30,
    format: 'vertical'
  },
  instagram: {
    size: '1080:1920',
    bitrate: '3500k',
    fps: 30,
    format: 'vertical'
  },
  youtube_shorts: {
    size: '1080:1920',
    bitrate: '5000k',
    fps: 60,
    format: 'vertical'
  },
  youtube: {
    size: '1920:1080',
    bitrate: '5000k',
    fps: 60,
    format: 'horizontal'
  }
};

/**
 * Crée un fichier de liste pour ffmpeg
 */
async function createFileList(videoFiles, outputDir) {
  const filelistPath = path.join(outputDir, 'filelist.txt');
  const content = videoFiles
    .map(file => `file '${path.resolve(file)}'`)
    .join('\n');

  await fs.writeFile(filelistPath, content);
  console.log(`📝 Fichier de liste créé: ${filelistPath}`);

  return filelistPath;
}

/**
 * Récupère tous les fichiers vidéo triés
 */
async function getVideoFiles(inputDir) {
  const files = await fs.readdir(inputDir);

  const videoFiles = files
    .filter(f => f.endsWith('.mp4') && f.startsWith('video_part_'))
    .sort()
    .map(f => path.join(inputDir, f));

  return videoFiles;
}

/**
 * Compile les vidéos avec ffmpeg
 */
async function compileVideos(inputDir, outputFile) {
  try {
    console.log('🎬 Récupération des vidéos...');
    const videoFiles = await getVideoFiles(inputDir);

    if (videoFiles.length === 0) {
      throw new Error('Aucune vidéo trouvée dans le dossier');
    }

    console.log(`📹 ${videoFiles.length} vidéos trouvées`);

    // Créer le fichier de liste
    const filelistPath = await createFileList(videoFiles, inputDir);

    // Compiler avec ffmpeg
    console.log('🔄 Compilation en cours...');
    const cmd = `ffmpeg -f concat -safe 0 -i "${filelistPath}" -c copy -y "${outputFile}"`;

    const { stdout, stderr } = await execAsync(cmd);

    // Vérifier la taille du fichier
    const stats = await fs.stat(outputFile);
    const sizeMB = (stats.size / (1024 * 1024)).toFixed(2);

    console.log(`✅ Vidéo compilée avec succès: ${outputFile}`);
    console.log(`📊 Taille: ${sizeMB} MB`);

    return true;
  } catch (error) {
    console.error('❌ Erreur lors de la compilation:', error.message);
    return false;
  }
}

/**
 * Optimise la vidéo pour une plateforme spécifique
 */
async function optimizeForPlatform(inputFile, outputFile, platform = 'tiktok') {
  try {
    const config = PLATFORM_CONFIGS[platform] || PLATFORM_CONFIGS.tiktok;

    console.log(`⚙️ Optimisation pour ${platform}...`);

    const cmd = `ffmpeg -i "${inputFile}" \
      -vf "scale=${config.size}:force_original_aspect_ratio=decrease,pad=${config.size}:(ow-iw)/2:(oh-ih)/2" \
      -r ${config.fps} \
      -b:v ${config.bitrate} \
      -c:v libx264 \
      -preset medium \
      -c:a aac \
      -b:a 128k \
      -y "${outputFile}"`;

    await execAsync(cmd);

    console.log(`✅ Optimisation pour ${platform} terminée`);
    return true;
  } catch (error) {
    console.error('❌ Erreur lors de l\'optimisation:', error.message);
    return false;
  }
}

/**
 * Ajoute un filigrane texte
 */
async function addTextWatermark(inputFile, outputFile, text, position = 'bottom-right') {
  try {
    const positions = {
      'top-left': 'x=10:y=10',
      'top-right': 'x=W-tw-10:y=10',
      'bottom-left': 'x=10:y=H-th-10',
      'bottom-right': 'x=W-tw-10:y=H-th-10',
      'center': 'x=(W-tw)/2:y=(H-th)/2'
    };

    const pos = positions[position] || positions['bottom-right'];

    console.log('🏷️ Ajout du filigrane texte...');

    const cmd = `ffmpeg -i "${inputFile}" \
      -vf "drawtext=text='${text}':fontsize=24:fontcolor=white@0.7:${pos}" \
      -codec:a copy \
      -y "${outputFile}"`;

    await execAsync(cmd);

    console.log('✅ Filigrane ajouté');
    return true;
  } catch (error) {
    console.error('❌ Erreur lors de l\'ajout du filigrane:', error.message);
    return false;
  }
}

/**
 * Ajoute une intro et/ou outro
 */
async function addIntroOutro(inputFile, outputFile, intro = null, outro = null) {
  try {
    const videos = [];

    if (intro && await fs.access(intro).then(() => true).catch(() => false)) {
      videos.push(intro);
    }

    videos.push(inputFile);

    if (outro && await fs.access(outro).then(() => true).catch(() => false)) {
      videos.push(outro);
    }

    if (videos.length === 1) {
      console.log('ℹ️ Pas d\'intro/outro à ajouter');
      return true;
    }

    console.log('🎥 Ajout intro/outro...');

    const tempDir = path.dirname(inputFile);
    const filelistPath = await createFileList(videos, tempDir);

    const cmd = `ffmpeg -f concat -safe 0 -i "${filelistPath}" -c copy -y "${outputFile}"`;
    await execAsync(cmd);

    console.log('✅ Intro/outro ajoutées');
    return true;
  } catch (error) {
    console.error('❌ Erreur lors de l\'ajout intro/outro:', error.message);
    return false;
  }
}

/**
 * Ajoute une musique de fond
 */
async function addBackgroundMusic(inputFile, outputFile, musicFile, volume = 0.3) {
  try {
    console.log('🎵 Ajout de la musique de fond...');

    const cmd = `ffmpeg -i "${inputFile}" -i "${musicFile}" \
      -filter_complex "[1:a]volume=${volume}[a1];[0:a][a1]amix=inputs=2:duration=first[aout]" \
      -map 0:v -map "[aout]" \
      -c:v copy \
      -c:a aac \
      -y "${outputFile}"`;

    await execAsync(cmd);

    console.log('✅ Musique de fond ajoutée');
    return true;
  } catch (error) {
    console.error('❌ Erreur lors de l\'ajout de la musique:', error.message);
    return false;
  }
}

/**
 * Génère des sous-titres automatiques (nécessite whisper ou service API)
 */
async function generateSubtitles(videoFile, outputSrtFile) {
  try {
    console.log('📝 Génération des sous-titres...');

    // Extraire l'audio
    const audioFile = videoFile.replace('.mp4', '.wav');
    await execAsync(`ffmpeg -i "${videoFile}" -vn -acodec pcm_s16le -ar 16000 -ac 1 "${audioFile}"`);

    console.log('ℹ️ Audio extrait. Utilisez un service comme Whisper API pour générer les sous-titres.');
    console.log(`   Audio: ${audioFile}`);
    console.log(`   Output SRT: ${outputSrtFile}`);

    // NOTE: Ici, vous devriez intégrer avec un service de transcription
    // comme OpenAI Whisper, Google Speech-to-Text, etc.

    return true;
  } catch (error) {
    console.error('❌ Erreur lors de la génération des sous-titres:', error.message);
    return false;
  }
}

/**
 * Fonction principale
 */
async function main() {
  const args = process.argv.slice(2);

  if (args.length < 2) {
    console.log('Usage: node compile_videos.js <input_dir> <output_file> [options]');
    console.log('\nOptions:');
    console.log('  --platform <platform>     Optimiser pour: tiktok, instagram, youtube_shorts, youtube');
    console.log('  --watermark <text>        Ajouter un filigrane texte');
    console.log('  --intro <file>            Ajouter une vidéo d\'intro');
    console.log('  --outro <file>            Ajouter une vidéo d\'outro');
    console.log('  --music <file>            Ajouter une musique de fond');
    console.log('  --music-volume <0-1>      Volume de la musique (défaut: 0.3)');
    process.exit(1);
  }

  const inputDir = args[0];
  const outputFile = args[1];

  // Parser les options
  const options = {
    platform: null,
    watermark: null,
    intro: null,
    outro: null,
    music: null,
    musicVolume: 0.3
  };

  for (let i = 2; i < args.length; i += 2) {
    const flag = args[i];
    const value = args[i + 1];

    switch (flag) {
      case '--platform':
        options.platform = value;
        break;
      case '--watermark':
        options.watermark = value;
        break;
      case '--intro':
        options.intro = value;
        break;
      case '--outro':
        options.outro = value;
        break;
      case '--music':
        options.music = value;
        break;
      case '--music-volume':
        options.musicVolume = parseFloat(value);
        break;
    }
  }

  console.log('🚀 Début de la compilation des vidéos');
  console.log(`📂 Dossier d'entrée: ${inputDir}`);
  console.log(`📹 Fichier de sortie: ${outputFile}`);

  // Fichiers temporaires
  const tempDir = path.dirname(outputFile);
  const tempFile1 = path.join(tempDir, 'temp_compiled.mp4');
  const tempFile2 = path.join(tempDir, 'temp_optimized.mp4');
  const tempFile3 = path.join(tempDir, 'temp_watermark.mp4');

  try {
    // 1. Compilation de base
    if (!await compileVideos(inputDir, tempFile1)) {
      process.exit(1);
    }

    let currentFile = tempFile1;

    // 2. Ajouter intro/outro
    if (options.intro || options.outro) {
      await addIntroOutro(currentFile, tempFile2, options.intro, options.outro);
      currentFile = tempFile2;
    }

    // 3. Ajouter musique de fond
    if (options.music) {
      const musicOutput = tempFile3;
      await addBackgroundMusic(currentFile, musicOutput, options.music, options.musicVolume);
      currentFile = musicOutput;
    }

    // 4. Ajouter filigrane
    if (options.watermark) {
      const watermarkOutput = path.join(tempDir, 'temp_final.mp4');
      await addTextWatermark(currentFile, watermarkOutput, options.watermark);
      currentFile = watermarkOutput;
    }

    // 5. Optimiser pour la plateforme
    if (options.platform) {
      await optimizeForPlatform(currentFile, outputFile, options.platform);
    } else {
      await fs.rename(currentFile, outputFile);
    }

    // Nettoyage des fichiers temporaires
    const tempFiles = [tempFile1, tempFile2, tempFile3, path.join(tempDir, 'temp_final.mp4')];
    for (const file of tempFiles) {
      try {
        await fs.unlink(file);
      } catch (e) {
        // Ignorer si le fichier n'existe pas
      }
    }

    console.log('✅ Compilation terminée avec succès!');
    console.log(`📊 Durée approximative: ${8 * 8} secondes (8 parties x 8 secondes)`);

  } catch (error) {
    console.error('❌ Erreur fatale:', error.message);
    process.exit(1);
  }
}

// Exporter les fonctions pour utilisation dans n8n
module.exports = {
  compileVideos,
  optimizeForPlatform,
  addTextWatermark,
  addIntroOutro,
  addBackgroundMusic,
  generateSubtitles,
  PLATFORM_CONFIGS
};

// Exécuter si appelé directement
if (require.main === module) {
  main();
}

File: scripts/compile_videos.py
#!/usr/bin/env python3
"""
Script de compilation de vidéos pour l'automatisation n8n
Compile 8 vidéos de 8 secondes en une seule vidéo
"""

import os
import subprocess
import sys
import json
from pathlib import Path
from typing import List


def create_filelist(video_files: List[str], output_path: str) -> str:
    """
    Crée un fichier texte listant toutes les vidéos à concaténer

    Args:
        video_files: Liste des chemins des fichiers vidéo
        output_path: Chemin du dossier de sortie

    Returns:
        Chemin du fichier de liste créé
    """
    filelist_path = os.path.join(output_path, 'filelist.txt')

    with open(filelist_path, 'w') as f:
        for video in sorted(video_files):
            # Format requis par ffmpeg
            f.write(f"file '{os.path.abspath(video)}'\n")

    return filelist_path


def compile_videos(input_dir: str, output_file: str) -> bool:
    """
    Compile toutes les vidéos du dossier en une seule vidéo

    Args:
        input_dir: Dossier contenant les vidéos à compiler
        output_file: Chemin du fichier de sortie

    Returns:
        True si la compilation a réussi, False sinon
    """
    try:
        # Récupérer tous les fichiers MP4
        video_files = sorted([
            os.path.join(input_dir, f)
            for f in os.listdir(input_dir)
            if f.endswith('.mp4') and f.startswith('video_part_')
        ])

        if not video_files:
            print("❌ Aucune vidéo trouvée dans le dossier")
            return False

        print(f"📹 {len(video_files)} vidéos trouvées")

        # Créer le fichier de liste
        filelist_path = create_filelist(video_files, input_dir)
        print(f"📝 Fichier de liste créé: {filelist_path}")

        # Compiler avec ffmpeg
        cmd = [
            'ffmpeg',
            '-f', 'concat',
            '-safe', '0',
            '-i', filelist_path,
            '-c', 'copy',
            '-y',  # Écraser si existe
            output_file
        ]

        print(f"🎬 Compilation en cours...")
        result = subprocess.run(cmd, capture_output=True, text=True)

        if result.returncode == 0:
            file_size = os.path.getsize(output_file) / (1024 * 1024)  # MB
            print(f"✅ Vidéo compilée avec succès: {output_file}")
            print(f"📊 Taille: {file_size:.2f} MB")
            return True
        else:
            print(f"❌ Erreur lors de la compilation:")
            print(result.stderr)
            return False

    except Exception as e:
        print(f"❌ Erreur: {str(e)}")
        return False


def add_intro_outro(input_file: str, output_file: str, intro: str = None, outro: str = None) -> bool:
    """
    Ajoute une intro et/ou outro à la vidéo compilée

    Args:
        input_file: Vidéo d'entrée
        output_file: Vidéo de sortie
        intro: Chemin de la vidéo d'intro (optionnel)
        outro: Chemin de la vidéo d'outro (optionnel)

    Returns:
        True si réussi, False sinon
    """
    try:
        videos = []
        if intro and os.path.exists(intro):
            videos.append(intro)

        videos.append(input_file)

        if outro and os.path.exists(outro):
            videos.append(outro)

        if len(videos) == 1:
            print("ℹ️ Pas d'intro/outro à ajouter")
            return True

        # Créer un fichier de liste temporaire
        temp_dir = os.path.dirname(input_file)
        filelist = create_filelist(videos, temp_dir)

        cmd = [
            'ffmpeg',
            '-f', 'concat',
            '-safe', '0',
            '-i', filelist,
            '-c', 'copy',
            '-y',
            output_file
        ]

        print(f"🎥 Ajout intro/outro...")
        result = subprocess.run(cmd, capture_output=True, text=True)

        return result.returncode == 0

    except Exception as e:
        print(f"❌ Erreur lors de l'ajout intro/outro: {str(e)}")
        return False


def add_watermark(input_file: str, output_file: str, watermark_text: str = None, watermark_image: str = None) -> bool:
    """
    Ajoute un filigrane (texte ou image) à la vidéo

    Args:
        input_file: Vidéo d'entrée
        output_file: Vidéo de sortie
        watermark_text: Texte du filigrane (optionnel)
        watermark_image: Chemin de l'image de filigrane (optionnel)

    Returns:
        True si réussi, False sinon
    """
    try:
        if watermark_text:
            # Ajouter un filigrane texte
            cmd = [
                'ffmpeg',
                '-i', input_file,
                '-vf', f"drawtext=text='{watermark_text}':fontsize=24:fontcolor=white@0.7:x=10:y=H-th-10",
                '-codec:a', 'copy',
                '-y',
                output_file
            ]
        elif watermark_image and os.path.exists(watermark_image):
            # Ajouter un filigrane image
            cmd = [
                'ffmpeg',
                '-i', input_file,
                '-i', watermark_image,
                '-filter_complex', 'overlay=W-w-10:H-h-10',
                '-codec:a', 'copy',
                '-y',
                output_file
            ]
        else:
            print("ℹ️ Pas de filigrane à ajouter")
            return True

        print(f"🏷️ Ajout du filigrane...")
        result = subprocess.run(cmd, capture_output=True, text=True)

        return result.returncode == 0

    except Exception as e:
        print(f"❌ Erreur lors de l'ajout du filigrane: {str(e)}")
        return False


def optimize_for_platform(input_file: str, output_file: str, platform: str = 'tiktok') -> bool:
    """
    Optimise la vidéo pour une plateforme spécifique

    Args:
        input_file: Vidéo d'entrée
        output_file: Vidéo de sortie
        platform: Plateforme cible (tiktok, instagram, youtube)

    Returns:
        True si réussi, False sinon
    """
    try:
        # Configuration par plateforme
        configs = {
            'tiktok': {
                'size': '1080:1920',  # Format vertical 9:16
                'bitrate': '3000k',
                'fps': 30
            },
            'instagram': {
                'size': '1080:1920',  # Reels format
                'bitrate': '3500k',
                'fps': 30
            },
            'youtube': {
                'size': '1920:1080',  # Format horizontal 16:9 ou shorts 9:16
                'bitrate': '5000k',
                'fps': 60
            }
        }

        config = configs.get(platform.lower(), configs['tiktok'])

        cmd = [
            'ffmpeg',
            '-i', input_file,
            '-vf', f"scale={config['size']}:force_original_aspect_ratio=decrease,pad={config['size']}:(ow-iw)/2:(oh-ih)/2",
            '-r', str(config['fps']),
            '-b:v', config['bitrate'],
            '-c:v', 'libx264',
            '-preset', 'medium',
            '-c:a', 'aac',
            '-b:a', '128k',
            '-y',
            output_file
        ]

        print(f"⚙️ Optimisation pour {platform}...")
        result = subprocess.run(cmd, capture_output=True, text=True)

        return result.returncode == 0

    except Exception as e:
        print(f"❌ Erreur lors de l'optimisation: {str(e)}")
        return False


def main():
    """Fonction principale"""
    if len(sys.argv) < 3:
        print("Usage: python compile_videos.py <input_dir> <output_file> [--platform tiktok|instagram|youtube]")
        sys.exit(1)

    input_dir = sys.argv[1]
    output_file = sys.argv[2]
    platform = None

    # Parser les arguments optionnels
    if '--platform' in sys.argv:
        idx = sys.argv.index('--platform')
        if idx + 1 < len(sys.argv):
            platform = sys.argv[idx + 1]

    print("🚀 Début de la compilation des vidéos")
    print(f"📂 Dossier d'entrée: {input_dir}")
    print(f"📹 Fichier de sortie: {output_file}")

    # Compilation basique
    temp_file = output_file.replace('.mp4', '_temp.mp4')
    if not compile_videos(input_dir, temp_file):
        sys.exit(1)

    # Optimisation pour la plateforme si spécifié
    if platform:
        if not optimize_for_platform(temp_file, output_file, platform):
            print("⚠️ Optimisation échouée, utilisation de la vidéo non optimisée")
            os.rename(temp_file, output_file)
        else:
            os.remove(temp_file)
    else:
        os.rename(temp_file, output_file)

    print("✅ Compilation terminée avec succès!")
    print(f"📊 Durée approximative: 64 secondes (8 parties x 8 secondes)")


if __name__ == '__main__':
    main()
